File: NextPermutationLC31.java
class Solution {

    //Time Complexity: O(n)
    //Space Complexity : O(1)
    
    public void nextPermutation(int[] nums) {
        
        if(nums == null || nums.length == 0) return;

        int n = nums.length;
        int i = n-2;

        //iterating through array starting from the end till we get the element which is not greater than it's right element
        while(i>=0 && nums[i] >= nums[i+1]){
            i--;
        }

        //if i>=0, if the whole array is in decending order then in that case i would be -1
        if(i>=0){
            
            int j = n-1;
            
            //comapering the ith element to it's right hand side numbers, starting from the end; whenever we get the number which is just greater than ith element, then we just swap those 2 numbers
            while(nums[j] <= nums[i]){
                j--;
            }
            swap(nums, i, j);
        }
        //reversing the array
        reverse(nums, i+1, n-1);
    }

    private void reverse(int[] nums, int start, int end){

        //reversing the array from start to end
        while(start<end){
            swap(nums, start, end);
            start++;
            end--;
        }
    }

    private void swap(int[] nums, int index1, int index2){
        
        //swapping the two numbers
        int temp = nums[index1];
        nums[index1] = nums[index2];
        nums[index2] = temp;
    }


    
    //Time Complexity: O(n)
    //Space Complexity : O(1)
    
    // public void nextPermutation(int[] nums) {
        
    //     int n = nums.length;
        
    //     for(int i= n-2; i>=0; i--){
            
    //         if(nums[i]< nums[i+1]){
                
    //             int j=n-1;
                
    //             while(j>i){
                    
    //                 if(nums[j]>nums[i]){
    //                     break;
    //                 }
    //                 j--;
    //             }
                
    //             swap(nums, i, j);
                
    //             reverse(nums, i+1, n-1);
                
    //             return;
                
    //         }
    //     }
    //     reverse(nums, 0, n-1);
    // }
    
    // private void swap(int[] nums, int i, int j){
        
    //     int temp = nums[i];
    //     nums[i] = nums[j];
    //     nums[j] = temp;   
    // }
    
    // private void reverse(int[] nums, int left, int right){
        
    //     while(left < right){
            
    //         swap(nums, left, right);
    //         left++;
    //         right--;
            
    //     }
    // }
}

File: next-permutation_LC_31.py
#TC:O(N) SC:O(1)
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        
        #start from right side, check with one previous if previous less than the curr check next greater than previous and swap.
        n=len(nums)-2
        j=len(nums)-1

        while(n>=0 and nums[n]>=nums[n+1]):
            n=n-1
        
        i=len(nums)-1
        while(n>=0 and i>=0 and nums[n]>=nums[i]): 
            i=i-1
       
        self.swap(nums,n,i)

        while(n+1<=j):
            self.swap(nums,n+1,j)
            n=n+1
            j=j-1

    def swap(self,nums, num1, num2):
        x=nums[num1] 
        nums[num1]=nums[num2]  
        nums[num2]=x   
    

